File: sdks/javascript/tests/TaskirXClient.test.ts
/**
 * TaskirX Client Tests
 */

import TaskirXClient from '../src/client';
import { ClientConfig } from '../src/types';

describe('TaskirXClient', () => {
  let client: TaskirXClient;
  const mockConfig: ClientConfig = {
    apiUrl: 'http://localhost:3000',
    apiKey: 'test-api-key-123',
    debug: true,
  };

  beforeEach(() => {
    client = new TaskirXClient(mockConfig);
  });

  describe('Initialization', () => {
    it('should initialize with valid config', () => {
      expect(client).toBeDefined();
      expect(client.auth).toBeDefined();
      expect(client.campaigns).toBeDefined();
      expect(client.analytics).toBeDefined();
      expect(client.bidding).toBeDefined();
      expect(client.ads).toBeDefined();
      expect(client.webhooks).toBeDefined();
    });

    it('should have all services available', () => {
      expect(client.auth).toBeTruthy();
      expect(client.campaigns).toBeTruthy();
      expect(client.analytics).toBeTruthy();
      expect(client.bidding).toBeTruthy();
      expect(client.ads).toBeTruthy();
      expect(client.webhooks).toBeTruthy();
    });

    it('should support debug mode toggling', () => {
      expect(() => client.enableDebug(true)).not.toThrow();
      expect(() => client.enableDebug(false)).not.toThrow();
    });

    it('should allow setting API key', () => {
      expect(() => client.setApiKey('new-api-key')).not.toThrow();
    });
  });

  describe('Service Access', () => {
    it('should access auth service', () => {
      expect(client.auth).toBeDefined();
      expect(typeof client.auth.register).toBe('function');
      expect(typeof client.auth.login).toBe('function');
      expect(typeof client.auth.logout).toBe('function');
    });

    it('should access campaigns service', () => {
      expect(client.campaigns).toBeDefined();
      expect(typeof client.campaigns.create).toBe('function');
      expect(typeof client.campaigns.list).toBe('function');
      expect(typeof client.campaigns.get).toBe('function');
      expect(typeof client.campaigns.update).toBe('function');
      expect(typeof client.campaigns.delete).toBe('function');
    });

    it('should access analytics service', () => {
      expect(client.analytics).toBeDefined();
      expect(typeof client.analytics.getRealtime).toBe('function');
      expect(typeof client.analytics.getCampaignAnalytics).toBe('function');
      expect(typeof client.analytics.getBreakdown).toBe('function');
    });

    it('should access bidding service', () => {
      expect(client.bidding).toBeDefined();
      expect(typeof client.bidding.submitBid).toBe('function');
      expect(typeof client.bidding.getRecommendations).toBe('function');
      expect(typeof client.bidding.getStats).toBe('function');
    });

    it('should access ads service', () => {
      expect(client.ads).toBeDefined();
      expect(typeof client.ads.create).toBe('function');
      expect(typeof client.ads.list).toBe('function');
      expect(typeof client.ads.delete).toBe('function');
    });

    it('should access webhooks service', () => {
      expect(client.webhooks).toBeDefined();
      expect(typeof client.webhooks.subscribe).toBe('function');
      expect(typeof client.webhooks.list).toBe('function');
      expect(typeof client.webhooks.onEvent).toBe('function');
    });
  });

  describe('Client Methods', () => {
    it('should have getProfile method', () => {
      expect(typeof client.getProfile).toBe('function');
    });

    it('should have logout method', () => {
      expect(typeof client.logout).toBe('function');
    });

    it('should have getDashboard method', () => {
      expect(typeof client.getDashboard).toBe('function');
    });

    it('should have getCampaignPerformance method', () => {
      expect(typeof client.getCampaignPerformance).toBe('function');
    });

    it('should have getStatistics method', () => {
      expect(typeof client.getStatistics).toBe('function');
    });

    it('should have createCampaigns batch method', () => {
      expect(typeof client.createCampaigns).toBe('function');
    });
  });

  describe('Configuration', () => {
    it('should use provided configuration', () => {
      const customConfig: ClientConfig = {
        apiUrl: 'https://api.example.com',
        apiKey: 'custom-key',
        debug: false,
        timeout: 60000,
        retryAttempts: 5,
      };

      const customClient = new TaskirXClient(customConfig);
      expect(customClient).toBeDefined();
    });

    it('should support minimal configuration', () => {
      const minimalConfig: ClientConfig = {
        apiUrl: 'http://localhost:3000',
        apiKey: 'test-key',
      };

      const minimalClient = new TaskirXClient(minimalConfig);
      expect(minimalClient).toBeDefined();
    });
  });
});

File: sdks/android/src/test/java/com/taskir/sdk/TaskirXClientTest.kt
package com.taskir.sdk

import android.content.Context
import androidx.test.core.app.ApplicationProvider
import com.taskir.sdk.data.models.*
import com.taskir.sdk.network.RequestManager
import com.taskir.sdk.services.AuthService
import com.taskir.sdk.services.CampaignService
import io.mockk.coEvery
import io.mockk.mockk
import junit.framework.TestCase
import kotlinx.coroutines.runBlocking
import org.junit.Before
import org.junit.Test

/**
 * Unit tests for TaskirX Android SDK
 */
class TaskirXClientTest {

    private lateinit var context: Context
    private lateinit var client: TaskirXClient
    private lateinit var config: ClientConfig

    @Before
    fun setup() {
        context = ApplicationProvider.getApplicationContext()
        config = ClientConfig(
            apiUrl = "http://localhost:3000",
            apiKey = "test-api-key",
            debug = true,
            timeout = 30000L,
            retryAttempts = 3
        )
        client = TaskirXClient(context, config)
    }

    @Test
    fun testClientInitialization() {
        TestCase.assertNotNull(client)
        TestCase.assertNotNull(client.auth)
        TestCase.assertNotNull(client.campaigns)
        TestCase.assertNotNull(client.analytics)
        TestCase.assertNotNull(client.bidding)
        TestCase.assertNotNull(client.ads)
        TestCase.assertNotNull(client.webhooks)
    }

    @Test
    fun testServiceAccess() {
        TestCase.assertNotNull(client.auth)
        TestCase.assertNotNull(client.campaigns)
        TestCase.assertNotNull(client.analytics)
        TestCase.assertNotNull(client.bidding)
        TestCase.assertNotNull(client.ads)
        TestCase.assertNotNull(client.webhooks)
    }

    @Test
    fun testDebugModeToggle() {
        client.enableDebug(true)
        client.enableDebug(false)
        // No exceptions should be thrown
    }

    @Test
    fun testSetApiKey() {
        client.setApiKey("new-api-key")
        // API key should be updated
    }

    @Test
    fun testConfigurationSupport() {
        val customConfig = ClientConfig(
            apiUrl = "https://api.example.com",
            apiKey = "custom-key",
            debug = false,
            timeout = 60000L,
            retryAttempts = 5
        )
        val customClient = TaskirXClient(context, customConfig)
        TestCase.assertNotNull(customClient)
    }

    @Test
    fun testVersionConstant() {
        TestCase.assertNotNull(TaskirXClient.VERSION)
        TestCase.assertEquals("1.0.0", TaskirXClient.VERSION)
    }
}

/**
 * Unit tests for AuthService
 */
class AuthServiceTest {

    private lateinit var authService: AuthService
    private lateinit var mockRequestManager: RequestManager

    @Before
    fun setup() {
        mockRequestManager = mockk()
        authService = AuthService(mockRequestManager, true)
    }

    @Test
    fun testLoginSuccess() = runBlocking {
        val mockResponse = AuthResponse(
            token = "jwt-token-123",
            user = User(
                id = "user-123",
                email = "test@example.com",
                name = "Test User"
            )
        )

        coEvery {
            mockRequestManager.executeWithRetry(any())
        } returns mockResponse

        val result = authService.login("test@example.com", "password")

        TestCase.assertNotNull(result)
        TestCase.assertEquals("jwt-token-123", result.token)
    }

    @Test
    fun testLogout() = runBlocking {
        coEvery {
            mockRequestManager.executeWithRetry(any())
        } returns Unit

        authService.logout()
        // Should complete without exception
    }

    @Test
    fun testGetProfile() = runBlocking {
        val mockUser = User(
            id = "user-123",
            email = "test@example.com",
            name = "Test User"
        )

        coEvery {
            mockRequestManager.executeWithRetry(any())
        } returns mockUser

        val result = authService.getProfile()

        TestCase.assertNotNull(result)
        TestCase.assertEquals("test@example.com", result.email)
    }

    @Test
    fun testTokenRefresh() = runBlocking {
        val mockResponse = AuthResponse(
            token = "new-jwt-token-456"
        )

        coEvery {
            mockRequestManager.executeWithRetry(any())
        } returns mockResponse

        val result = authService.refreshToken()

        TestCase.assertNotNull(result)
        TestCase.assertEquals("new-jwt-token-456", result.token)
    }

    @Test
    fun testTokenManagement() = runBlocking {
        authService.setToken("test-token")
        val retrievedToken = authService.getToken()
        TestCase.assertEquals("test-token", retrievedToken)
    }
}

/**
 * Unit tests for CampaignService
 */
class CampaignServiceTest {

    private lateinit var campaignService: CampaignService
    private lateinit var mockRequestManager: RequestManager

    @Before
    fun setup() {
        mockRequestManager = mockk()
        campaignService = CampaignService(mockRequestManager, true)
    }

    @Test
    fun testCreateCampaign() = runBlocking {
        val mockCampaign = Campaign(
            id = "campaign-123",
            name = "Test Campaign",
            budget = 10000.0
        )

        coEvery {
            mockRequestManager.executeWithRetry(any())
        } returns mockCampaign

        val result = campaignService.create(mockCampaign)

        TestCase.assertNotNull(result)
        TestCase.assertEquals("campaign-123", result.id)
    }

    @Test
    fun testListCampaigns() = runBlocking {
        val mockCampaigns = listOf(
            Campaign(id = "c1", name = "Campaign 1", budget = 5000.0),
            Campaign(id = "c2", name = "Campaign 2", budget = 10000.0)
        )

        coEvery {
            mockRequestManager.executeWithRetry(any())
        } returns mockCampaigns

        val result = campaignService.list()

        TestCase.assertNotNull(result)
        TestCase.assertEquals(2, result.size)
    }

    @Test
    fun testGetCampaign() = runBlocking {
        val mockCampaign = Campaign(
            id = "campaign-123",
            name = "Test Campaign",
            budget = 10000.0
        )

        coEvery {
            mockRequestManager.executeWithRetry(any())
        } returns mockCampaign

        val result = campaignService.get("campaign-123")

        TestCase.assertNotNull(result)
        TestCase.assertEquals("Test Campaign", result.name)
    }

    @Test
    fun testUpdateCampaign() = runBlocking {
        val updatedCampaign = Campaign(
            id = "campaign-123",
            name = "Updated Campaign",
            budget = 15000.0
        )

        coEvery {
            mockRequestManager.executeWithRetry(any())
        } returns updatedCampaign

        val result = campaignService.update("campaign-123", mapOf("budget" to 15000.0))

        TestCase.assertNotNull(result)
        TestCase.assertEquals(15000.0, result.budget)
    }

    @Test
    fun testDeleteCampaign() = runBlocking {
        coEvery {
            mockRequestManager.executeWithRetry(any())
        } returns Unit

        campaignService.delete("campaign-123")
        // Should complete without exception
    }

    @Test
    fun testPauseCampaign() = runBlocking {
        val pausedCampaign = Campaign(
            id = "campaign-123",
            name = "Test Campaign",
            budget = 10000.0,
            status = "paused"
        )

        coEvery {
            mockRequestManager.executeWithRetry(any())
        } returns pausedCampaign

        val result = campaignService.pause("campaign-123")

        TestCase.assertNotNull(result)
        TestCase.assertEquals("paused", result.status)
    }

    @Test
    fun testResumeCampaign() = runBlocking {
        val resumedCampaign = Campaign(
            id = "campaign-123",
            name = "Test Campaign",
            budget = 10000.0,
            status = "active"
        )

        coEvery {
            mockRequestManager.executeWithRetry(any())
        } returns resumedCampaign

        val result = campaignService.resume("campaign-123")

        TestCase.assertNotNull(result)
        TestCase.assertEquals("active", result.status)
    }
}

/**
 * Integration tests for common workflows
 */
class TaskirXIntegrationTest {

    private lateinit var context: Context
    private lateinit var client: TaskirXClient

    @Before
    fun setup() {
        context = ApplicationProvider.getApplicationContext()
        val config = ClientConfig(
            apiUrl = "http://localhost:3000",
            apiKey = "test-api-key"
        )
        client = TaskirXClient(context, config)
    }

    @Test
    fun testServiceConfiguration() {
        TestCase.assertNotNull(client.auth)
        TestCase.assertNotNull(client.campaigns)
        TestCase.assertNotNull(client.analytics)
        TestCase.assertNotNull(client.bidding)
        TestCase.assertNotNull(client.ads)
        TestCase.assertNotNull(client.webhooks)
    }

    @Test
    fun testErrorHandling() {
        // Error handling should be graceful
        TestCase.assertNotNull(client)
    }

    @Test
    fun testRetryLogic() {
        // Retry logic should be transparent
        TestCase.assertNotNull(client)
    }
}
